Propagate network errors through promise (issue 6139)

alt- opened the pull request:
Fixes #6139 

Propagates network errors through API promises so that they can be handled by the requester. Prevents rejected promises from being cached so requests can be retried.

Also adds support for testing connection errors in unit tests by exposing web server instantiation.

yurydelendik commented on src/core/chunked_stream.js:
```diff
@@ -329,12 +331,16 @@ var ChunkedStreamManager = (function ChunkedStreamManagerClosure() {
-      this.requestChunks(missingChunks);
-      return this._loadedStreamCapability.promise;
+      return new Promise(function(resolve, reject) {
+        this.requestChunks(missingChunks, function() {
```
is `this` here defined as ChunkedStreamManager?

alt- replied:
No, it's not. Good catch. Fixed

timvandermeij commented:
@alt- Could you squash the commits? See https://github.com/mozilla/pdf.js/wiki/Squashing-Commits on how to do that easily.

alt- committed:
Propagate network errors through promise

Propagates network errors through API promises so that they can be handled
by the requester. Prevents rejected promises from being cached so requests
can be retried.

alt- commented:
@timvandermeij Squashed

brendandahl commented:
Looks like there's been some code churn. Can you update and I'll review?

timvandermeij commented on package.json:
```diff
@@ -9,7 +9,8 @@
     "typogr": "~0.5.0",
-    "yargs": "~1.2.1"
+    "yargs": "~1.2.1",
+    "stream-throttle": "~0.1.3"
```
Nit: keep the list in alphabetical order by adding this one below ShellJS

brendandahl commented:
I haven't seen any activity on this for awhile, so I'm closing for now.
